\chapter{\poy Quick Start}

\section{Requirements: software and hardware}

\subsection{Software}
\poy is a platform-independent, open-source program that can be
compiled for many operating systems and hardware configurations,
including Mac OSX, Microsoft Windows and Linux. %\poy \emph{binaries}
%\index{general}{binaries} (compiled application file) is the only
piece of software necessary to run \poy%1.0.  The intuitive
\emph{Graphical User Interface} of \poy provides the functionality
for running analyses using pull-down menus and field selections,
as well as creating and running \poy scripts. Some utility programs
(such as Notepad and Ghostscript for Windows, TextEdit for Mac, or
Nano for Linux), can help preparing \poy scripts and formatting
data files, while others (such as Adobe Acrobat) can facilitate
viewing the graphical output in PDF (Portable Document Format).

\subsection{Hardware}
\poy runs on a variety of computers from laptops and desktops to
clusters of various sizes and symmetric multiprocessing hardware.
There are no particular requirements for disk space, but XML and
diagnosis report files can be large.

\section{Obtaining and installing \poy}
\subsection{Installing from the binaries}
\poy installers for Windows and Mac OSX, source code, and documentation
in PDF format are available from the \poy download website at the
Computational Sciences site of the American Museum of Natural
History:

\begin{center}
\url{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy}
\end{center}

The latest source code can also be obtained from \poy Google Group website:
\begin{center}
\url{http://code.google.com/p/poy/source}
\end{center}

The following detailed step-by-step instructions will guide you
through downloading and installing \poy \emph{binaries} for various
platforms.

\begin{flushleft}
\begin{minipage}[c]{0.074\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figlogowindows.jpg}
\end{minipage}
\,
\begin{minipage}[t]{0.88\textwidth}
\subsubsection*{Windows}
\end{minipage}


\begin{itemize}
\item
Download the
\href{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy/download}
{\emph poy5} folder to the desktop by selecting the \emph{Windows} 
download link.

\item Open the \emph{POY\_version.zip}.  You will need Administrator
privileges to install the application.  Extract the zip file to
install in the desired location and execute the \emph{poy.exe} file.
If you have Windows XP SP2, Windows Vista or Windows 7 and possess
more than one core or processor, you can take advantage of this
processing power \ by installing the parallel components
\href{http://www.mpich.org}{MPICH2}.

[Note: The \poy developers encountered no problems when using MPICH2 3.0.2.]
\end{itemize}

\begin{minipage}[c]{0.074\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figlogomac.jpg}
\end{minipage}
\,
\begin{minipage}[t]{0.88\textwidth}
\subsubsection*{Mac OSX}
\end{minipage}
\begin{itemize}
\item Download
\href{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy/download}
{\emph{poy-buildXXXX.dmg}} disk image to the desktop. The complete
installation of the Mac OSX version of \poy includes MPICH2 1.4.1,
which is used to communicate processes during parallel execution.
\item Drag the \poy application from the disk \emph{poy5} and drop
it into the \emph{Applications} folder on the hard drive.

[Note: During the first execution in parallel you may be asked by
the Firewall to unblock \poy and MPICH.  This is necessary for
successful execution of the program.] \end{itemize}

\begin{minipage}[c]{0.074\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figlogolinux.jpg}
\end{minipage}
\,
\begin{minipage}[t]{0.88\textwidth}
\subsubsection*{Linux}
\end{minipage}
\begin{itemize}
\item  No binaries are available for the Linux operating system.
The user should compile \poy directly from the source (see
Section~\ref{Compilingfromsource}).  \end{itemize}

\end{flushleft}

\subsection{Compiling from the source}
\label{Compilingfromsource}

For the majority of users, downloading the binaries from the \poy
download site will suffice.  However, in some cases it may be %either
necessary: analysis of chromosomal and genomic data (\textgreater16383
%nucleotides long); \hl{likelihood analyses}; or if working with
large alphabets (\textgreater255 elements), or% desirable --- user
preference for working in a command-line environment or running
\poy analyses in parallel (in the case of Linux machines or on a
cluster computer), or necessary (in the case of Linux users) ---
to compile \poy directly from the source code (see Table
\ref{InterfaceGuide} and \ref{ParallelizationGuide}).  If the user
chooses to compile, it is advisable to check out the various
configuration options that can be found in {\tt ./configure --help}
of the \texttt {src} directory. \\ \\ In order to compile \poy the
following tools are required:

\begin{enumerate}
\item The \href{http://www.gnu.org/software/make/}{GNU Make 3.8} utility. 
\item OCaml \href{http://www.ocaml.org}{version 3.11.2.} or later. 
\item A {\bf C} compiler, for example \href{http://gcc.gnu.org/} 
{The GNU Compiler Collection.}
\item \href{http://www.zlib.net}{The zlib compression library.}
\item The Linear Algebra PACKage \href {http://www.netlib.org/lapack/}
{LAPACK} must be installed in order to use the likelihood option.
\item The \href{http://www.gnu.org/s/ncurses} {ncurses library} 
is necessary to compile the default interface, i.e. \emph{ncurses} 
or the \emph{Interactive Console}. If this library is not available, 
the \emph{flat} interface will be compiled instead.
\item The Message Passing Interface \href{http://www-unix.mcs.anl.gov/mpi/}
{MPICH2}, which is used to communicate processes during parallel execution.
%  the complete installation includes MPICH2 \hl{1.06 p1}. MPICH is used to 
%communicate processes during parallel execution (if you didn't know what it 
%is you most likely don't have it), select the custom installation option 
%and remove that component. During the first execution in parallel you 
%will be asked by the Windows Firewall to %unblock \poy and MPICH: 
%this is necessary for the successful execution of the program.
\item The \href{http://www.tsp.gatech.edu/concorde/}{Concorde Package}, 
which includes TSP median solvers used in the calculation of 
chromosome and genome rearrangements. 
\end{enumerate}
Download, ungzip and untar the
\href{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy/download}
{\poy source code}.  In a terminal window, change directories to the path of this 
uncompressed directory. In order to compile under default setting, 
working in the source directory (\texttt{src}), type:

\begin{verbatim}
./configure
make
make install
\end{verbatim}

To \texttt{make install}, it may be necessary to do this as the
root user using \texttt{sudo}.  This script will compile the
\emph{Interactive Console} or \emph{ncurses} interface that will
be found in \commandstyle{/src/\_build}.  Another configuration
option includes a \emph{readline} interface.  Similar to the
\emph{ncurses} interface, this allows for the use of arrow keys to
modify commands and browse command history.  %An \emph{html} interface
is available.  A \emph{flat} interface is also available that
supports the running of the program in parallel, irrespective of
the operating system. This version is run through a terminal window
and invoked in a script (see Section~\ref{sec: ExecutingScript}).
In order to run \poy in parallel environments,
\href{http://www-unix.mcs.anl.gov/mpi/}{Message Passing Interface}
(invoked by \texttt{mpiexec} or \texttt{mpirun}, depending on your
implementation) must be invoked.  More than likely, your system
administrator already has one installed and should be able to provide
you with the proper paths to set your config file. In order to
compile this \emph{flat} version with parallel support type:

\begin{verbatim}
./configure --enable-interface=flat --enable-mpi CC=mpicc
make 
make install 
\end{verbatim} 

[Note: CC=mpicc is not available
for the Windows version of mpi, therefore it is not necessary to
include this component in the compiling script.]

Table \ref{InterfaceGuide} should be used as a guide as to the type of 
interface that should be employed depending on the type of data (`standard' 
or `long').

\begin{table}[th!]
\small
\caption{Interface Guide. `Standard' data equates to a molecular sequence 
or single partition that is fewer than 16383 nucleotides in length or contain 
fewer than 255 elements, while `long' data partitions accommodate lengths 
greater than these values. The field `config+' indicates that the options 
long-sequences and/or large-alphabets must be enabled during compilation 
for these datatypes to be analyzed. A distinction is made between the 
\emph{Interactive Console} or \emph{ncurses} that is downloaded as binaries 
(\texttt{bins}) from the website and that which is compiled from the source 
(\texttt{src}). [Note: It is not possible to analyze either long sequences or
large alphabets using the \emph{GUI} or the \emph{Interactive Console} 
when downloaded from the \poy website.]}
\label{InterfaceGuide} 
\begin{center}
\renewcommand{\arraystretch}{1.5}
\begin{tabular}{p{2.0cm}  p{0.75cm}  p{3.0cm}  p{3.0cm}  p{1.0cm}} 
\hline
Data type & GUI & Interactive Console (bins) & Interactive Console (src) & Flat \\
\hline
Standard & + & + & + & + \\
Long & -- & -- & config+ & config+\\
\hline
\end{tabular}
\end{center}
\end{table}

\section {Executing a Script}
\label {sec: ExecutingScript}
A number of startup options are available when executing a script through 
the command line in a terminal window. The options below, can be viewed 
by typing \texttt{poy -help} in a terminal window.\\
\\
\texttt{-w} \hspace{0.25 cm} Run poy in the specified working directory\\
\texttt{-e}  \hspace{0.25 cm} Exit upon error\\
\texttt{-d}  \hspace{0.25 cm} Dump filename in case of error\\
\texttt{-q}  \hspace{0.25 cm} Don't wait for input other than the program 
argument script\\
\texttt{-no-output-xml}  \hspace{0.25 cm} Do not generate the output.xml file\\
\texttt{-plugin}  \hspace{0.25 cm} Load the selected plugins\\
\texttt{-script}  \hspace{0.25 cm} Inlined script to be included in the analysis\\
\texttt{-help}  \hspace{0.25 cm} Display this list of options\\

The use of these options are appropriate any time the user chooses to execute 
\poy from the command line,  when working with \emph{ncurses} and 
\emph{flat} interfaces. These options are also useful when operating \poy 
in a cluster environment. For example typing

\begin{quote}
\commandstyle{poy -w /Users/username/poyfiles mol.poy}
\end{quote}
in a terminal window will invoke the program POY to run the script 
\texttt{mol.poy} in the directory \texttt{/Users/username/poyfiles}.  
This is the equivalent of including

\begin{quote}
\commandstyle{cd ("/Users/username/poyfiles")}\\
\end{quote}

When attempting to run \poy in parallel from the command line, the programs 
\href{http://www-unix.mcs.anl.gov/mpi/}{MPI} or \href {https://www.osc.edu/~djohnson/mpiexec/} 
{Mpiexec} must be used to initialize the parallel job from within a PBS batch 
or interactive environment.  For example typing

\begin{quote}
\commandstyle{mpirun -np 8 $\sim$/POY\_bins/poy5a-mpi $\sim$/POY\_analyses/mol.poy}\\
\end{quote}
in the terminal window will invoke a parallel version of \poy to run the 
script \texttt{mol.poy}, utilizing 8 processors (Figure~\ref{fig:mpiexecscript}).  
The file \texttt{mol.poy} resides in the directory \commandstyle{POY\_analyses}, 
which is found in the home directory.
This parallel version of \poy was compiled in the directory 
\texttt{POY\_bins} in the home directory.  With the flat interface, it is not 
possible to run parallel jobs interactively, therefore a script \textbf {\emph{must}} 
be included in the command, so that it can be passed to the application.  

\begin{figure}
\begin{center}
\includegraphics[width=0.9\textwidth]{doc/figures/mpiexec_script.jpg}
\end{center}
\caption{\poy flat interface displayed in a terminal window. The 
interface indicates that the program has been compiled with `parallel 
on'. The program is running the script \texttt{mol.poy} in parallel 
over 8 processors. In this case, MPICH is used to communicate 
processes during  parallel execution.}
\label{fig:mpiexecscript}
\end{figure}

%-----------------------------------
%The Graphical User Interface
%-----------------------------------

\section{The Graphical User Interface}

Two of the working environments that \poy provides are the
\emph{Graphical User Interface} and the \emph{Interactive Console}
(also known as the \emph{ncurses} interface). The \emph{Graphical
User Interface} has a user-friendly appearance like other native
stand-alone applications where different functions are accessible
through menus and windows. Thus, the entire analysis can be carried
out by clicking on appropriate selections and, where necessary,
typing specifications in designated fields. Currently, the
\emph{Graphical User Interface} is designed for the analysis of
data with parsimony or likelihood (with minimal model selection)
as the optimality criterion.  Unlike the \emph{Interactive Console},
it is not possible to specify all options with the \emph{Graphical
User Interface}.  The minimum screen size for the \emph{Graphical
User Interface} is 1024 x 768 pixels.

On the other hand, the \emph{Interactive Console} requires a detailed
knowledge of \poy commands, their arguments, and the conventions
of \poy scripting. All these features are described in the \emph{POY
Commands} chapter (\ref{commands}).

Even though the Mac OSX version of the \emph{Graphical User Interface}
is used for screen shots throughout this chapter, the Windows version
contains the same items and functionality, differing only in the
generic window format specific to the platform.

When \poy is first opened, two items appear on the screen: the \poy
menu bar across the top and the \emph{POY Launcher} window
(Figure~\ref{fig:menu_launcher_window}).

[Note: In Windows the menu bar is within the launcher window.]

\begin{figure}[htpb]
\begin{center}
\includegraphics[width=0.75\textwidth]{doc/figures/menu_launcher_window.jpg}
\end{center}
\caption{The \poy menu bar and the \emph{POY Launcher} window. 
These items appear when \poy is opened.}
\label{fig:menu_launcher_window}
\end{figure}

\subsection{POY menu bar}
The menu bar contains the following drop-down menus:
\begin{description}
\item[POY] (Mac OSX only) Contains generic items as with other 
Mac OSX applications. This pull down menu allows selection of the 
\emph{About POY} window (Figure~\ref{fig:about_window}) that 
lists the current version of \poy, a copyright statement, and the address 
of the \poy website. In addition, it includes a \emph{Quit POY} tab 
that closes the program. 
\item[Analyses]    Contains options for different types of tree searches, 
calculation of support values, tree diagnosis, and their respective outputs. 
Other items in this menu open the \emph{POY Launcher} 
(Open Launcher) and the \emph{Interactive Console}.
\item[Edit] Contains standard tools for undoing, cutting, copying, 
pasting, deleting, and selecting.
\item[View] opens the \emph{Output} window to display the results 
(including warning and error messages) and the current state of the analysis. 
This \emph{Output} window also contains an \emph{update} tab.  
It also contains the \emph{About POY} menu item in Windows. %and Linux.  
\item[Help] Opens the \poy \emph{Manual} in PDF format (requires 
a PDF viewer).
\end{description}

\begin{figure}[htpb]
\begin{center}
\includegraphics[width=0.5\textwidth]{doc/figures/about_window.jpg}
\end{center}
\caption{The \emph{About POY} window.}
\label{fig:about_window}

\end{figure}

\subsection{POY Launcher} 
The \emph{POY Launcher} is the only window that automatically 
opens upon starting \poy. This allows the user to import a previously 
created script, designate a working directory, specify the number of 
processors, and start the analysis.

\begin{description}
\item[Select the script to run] 
Allows the user to specify the location of a \poy script.
\item[Select the working directory] 
The working directory is the directory that contains 
the input data and output files. By default, the working directory is set to 
be the same as the directory containing the selected \poy script. 
\item[Select the number of processors] 
If more than one processor or core is available, up to 
8 can be designated for running the analysis. It is important to note 
that once specified, the selection is applied to \emph{all} subsequent 
analyses in the current \poy session. Table \ref{ParallelizationGuide} 
is a guide to the parallelization ability of \poy depending on the operating 
system and the \poy interface being used. Observe that parallelization 
is \textbf{\emph {never}} supported in interactive sessions, see 
Section~\ref{interactiveconsole}.
\item[Run the analysis] 
Clicking the \emph{Run} button starts the execution of the selected 
script. Once the script is initiated, the \emph{Run} button becomes 
the \emph{Cancel} button that can be used to interrupt a \poy session.
\end{description}

\begin{table}[t] 
\small
\caption{Parallelization Guide. The field `mpi+' indicates that mpi 
must be enabled during compiling. A distinction is made between the 
\emph{Interactive Console} or \emph{ncurses} that is downloaded as 
binaries (\texttt{bins}) from the website and that which is compiled 
from the source (\texttt{src}).} 
\label{ParallelizationGuide} 
\begin{center}
\renewcommand{\arraystretch}{1.5}
\begin{tabular}{p{2.7cm}  p{1.1cm}  p{3.0cm}  p{3.0cm}  p{0.75cm}} 
\hline
Operating System & GUI & Interactive Console (bins) & Interactive Console (src) & Flat \\
\hline
Mac OSX & + & -- & -- & mpi+ \\
Windows & + & -- & -- & mpi+ \\
Linux & N/A & -- & -- & mpi+ \\
\hline
\end{tabular}
\end{center}
\end{table}

If the \emph{Run} button is clicked without the selected script and
working directory, or the names of the scripts and working directory 
are entered incorrectly, \poy issues an error message in the upper part 
of the \emph{POY Launcher} window, such as \texttt{POY finished 
with an error}.

\subsection{The \emph{Analyses} menu}
The \emph{Analyses} menu is the main toolbox of the \poy \emph{GUI} 
interface (Figure~\ref{fig:simple_search_window}, left). Selections are 
subdivided into four functional categories. The first three deal with tree 
searching, support calculation, and tree diagnosis; the fourth one is used 
for  script management or interactive command execution that bypasses the 
menu-driven script generation. Each of the menu items is described below 
in the order it appears on the menu.

Most options are consistently applied through different kinds of analysis. 
Therefore, all options are described in detail only for the \emph{Simple Search} 
analysis. The descriptions of other analyses are made with reference to the the 
\emph{Simple Search} and focus on unique options.

%-----------------------------------
%Tree search options
%-----------------------------------

\subsubsection{Tree searching options}

A number of different tree searching options are available through
the \emph{Graphical User Interface}.  These include a \emph{Simple
Search}, \emph{Timed Search}, \emph{Search with Ratchet} and
\emph{Search with Perturb}.

\subsubsection*{\emph{Simple Search}} 
The \emph{Simple Search}
window permits the analysis of a number of different data types,
including a range of molecular characters (from DNA sequences to
complete genomes), custom alphabet characters, and qualitative
characters, under parsimony.  It is also possible to carry out a
likelihood analysis of DNA sequence, morphological and qualitative
data under a number of different models.  In the simplest sense, a
typical search involves a series of steps.  First, initial trees
are generated by random addition sequence from the imported character
data.  These trees are then subjected to branch swapping, after
which trees are selected to report.  The \emph{Simple Search} window
(Figure~\ref{fig:simple_search_window}, right) provides the most
common and basic options for a standard tree search in \poy that
must be selected by either clicking the appropriate buttons or by
typing. Note that \emph{all} the empty fields must be filled in
(even if that means assigning a cost of \texttt{0} to all the
\emph{Sequence Parameters}), otherwise the default values will be
used. The window is subdivided into five sections: 

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/simplesearch_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/simplesearch_window_filled.jpg}
\end{minipage}

\caption{The \emph{Simple Search} window. Selecting \emph{Simple Search} 
from the \emph{Analysis} 
menu (left) opens the \emph{Simple Search} window options (right).}
\label{fig:simple_search_window}
\end{figure}

\begin{description}
\setlength{\parindent}{0.5cm}
\item[Input Files]
Contains the list of files that are to be input into \poy. These include
character files in nucleotide, Hennig86, and Nexus formats, as well 
as tree files. Continuous characters can be input into \poy in a Hennig86 
format matrix (see~\ccross{read}). Character data in other formats can 
be input by specifying additional arguments in the script (see~\ccross{read}).

\begin{statement}
Gap-opening cost greater than $0$ can not be specified with 
prealigned data as the columns are treated as independent in 
this file format.
\end{statement}

\item[Search Parameters]
Holds one field to set the number of independent random addition Wagner 
replicates to be generated.

\item[Input Parameters]
Holds fields to specify the optimality criterion (parsimony or likelihood).  
With \emph{Parsimony} as the optimality criterion, it is possible to select 
different datatypes (sequence, chromosome, genome, custom alphabet, 
break inversion or qualitative) and allows the user to select whether these data 
should be treated as prealigned (if possible). Selection of different datatypes 
will invoke an additional subsection (see below).  Currently, it is only possible 
to analysis sequence and qualitative data with the maximum likelihood criterion.
\end{description}   

\hangindent=1cm	\emph{Parsimony Optimality Criterion}
The parameters in this section are dependent on the data types selected 
in \emph{Input Parameters}. More detailed explanations of the different 
data types can be found below and  in the difference character types 
sections of both~\ccross{read} and ~\ccross{transform}.

\begin{description}
%\setlength{\labelwidth}{0pt}
\setlength{\labelsep}{5pt}
\setlength{\itemindent}{0pt}%
\setlength{\parindent}{0.5cm}        

\item [Sequence Parameters]  If \emph{sequence} data types are
chosen, the user can specify the substitution, indel, and gap opening
costs of sequences. Enter \texttt{0} if no gap opening cost is
desired. If the value of a parameter is not specified, default
values are used.  The default value for both substitutions and
insertion deletion events is \texttt{1} and that of gap-opening is
\texttt{0}.

\item [Chromosome and Mauve Parameters] \emph{Chromosome} characters
are multi-locus nucleotide sequences and can include nuclear
chromosomes, as well as, mitochondrial and viral genomes.  It is
possible to submit either \poyargument{annotated} (by selection of
the \emph{Annotated} box) or \emph{Unannotated} (by leaving the
\emph{Annotated} box unchecked) chromosomes. Within \emph{Annotated}
chromosomes, homologous regions, such as loci, are separated with
the pipe symbol (``$\vert$'').  \emph{Unannotated} chromosomes are
entirely without delimiters. For \emph{Unannotated} chromosomes,
the Mauve Parameters, must be set by the user.  These parameters
(\emph{Match Quality, Match Coverage, Min. Match} and \emph {Max.
Match}) are employed by the Mauve aligner to find regions of homology
or synteny blocks between chromosomes (see~\nccross{annotate}{annotate}
within the command \poycommand{transform}). Default values for these
parameters are provided in the \emph{GUI}.

\indent Within this subsection it is necessary to specify both
\emph{Locus Indel} and rearrangement (\emph{Locus Breakpoint} or
\emph{Locus Inversion}) costs.  The cost of a \emph{Locus Indel}
is by default set to 10 plus 0.9 times the length of the locus
(see~\nccross{locus\_indel} {chromosomelocusindel}  within the
command \poycommand{transform}).  Rearrangements of homologous
regions---as defined by the user in the case of \emph{Annotated}
chromosomes or as determined by the Mauve aligner as in \emph{Unannotated}
chromosomes)---are then optimized using either \emph{Locus Breakpoint}
or \emph{Locus Inversion} costs
(see~\nccross{locus\_breakpoint}{chromosomelocusbreakpoint} and
~\nccross{locus\_inversion}{chromosomelocusinversion} within the
command \poycommand{transform}).  The default cost for both is set
to 10. The user must also specify the \emph{Median solver} for the
optimization of rearrangements of \emph{Annotated} chromosomes.
The default median solver is \emph{Caprara}, but the user can
alternatively choose \emph{BBTSP, ChainedLK, COALESTSP, MGR,
Siepel, SimpleLK} and \emph{Vinh}
(see~\nccross{median\_solver}{chromosomemediansolver} within the
command \poycommand{transform}).

\indent The user must also specify whether the chromosome is
\emph{Circular} (true) or linear (false).  It is not possible to
submit pre-aligned data files for either \emph{Annotated} or
\emph{Unannotated} chromosomes.


\begin{statement}
\textbf{Locus definition}. In the \emph{Sequence Parameters} section
for a parsimony analysis, the user may be required to specify the
cost associated with a \emph{Locus Breakpoint}, \emph{Locus Inversion}
or \emph{Locus Indel}, depending on the data type. In these cases,
\emph{Locus} should \textbf{not} be taken to be the functional,
biological unit in the classical sense, but only as a homologous
segment of a sequence.  \end{statement}

\item [Genome and Mauve Parameters] \emph{Genome} characters are
multi- \\locus, multi-chromosomal nucleotide sequences, wherein
transformations (i.e. indels, substitutions, and rearrangements)
are optimized at the sequence, locus \emph{and} chromosomal level.
Within the genome data file, individual chromosomes are separated
by the at symbol (``$@$'') and the individual chromosomes remain
\emph{Unannotated}.

\indent As with \emph{Unannotated Chromosome} characters, homologous
regions are determined using the Mauve parameters (\emph{Match
Quality, Match Coverage, Min. Match} and \emph {Max. Match}).
Default values for these parameters are provided in the \emph{GUI}.
The \emph{Locus Indel} and rearrangement (\emph{Locus Breakpoint}
or \emph{Locus Inversion}) costs are set by the user. By default,
the cost of a \emph{Locus Indel} is  set to 10 plus 0.9 times the
length of the locus (see the argument~\nccross{locus\_indel}
{chromosomelocusindel} within the \emph{Chromosome and genome
transformation methods} of the command \poycommand{transform}).
Rearrangements of homologous regions, as determined by the Mauve
aligner, are then optimized using either \emph{Locus Breakpoint}
or \emph{Locus Inversion} costs  (see the
arguments~\nccross{locus\_breakpoint}{chromosomelocusbreakpoint}
and~\nccross{locus\_inversion}{chromosomelocusinversion} within the
\emph{Chromosome and genome transformation methods} of the command
\poycommand{transform}).  The default cost for both is set to 10.
A \emph{Median solver} must also be specified for the optimization
of rearrangements: \emph{BBTSP, ChainedLK, COALESTSP, MGR, Siepel,
SimpleLK} and \emph{Vinh} (see the argument~\nccross{median\_solver}
{chromosomemediansolver} within the \emph{Chromosome and genome
transformation methods} of the command \poycommand{transform}).

\indent Two other costs must be set for the analysis of this data
type--\emph{Translocation} events and \emph{Chromosome Indel}. The
cost of the \emph{Translocation} of a region of one chromosome to
another chromosome is set to 10 by default.  The cost of the insertion
or deletion of an entire chromosome is by default set to 10 plus
0.9 times the length of the chromosome.

As with \emph{Chromosome} characters, it is not possible to input 
pre-aligned data files. 

\item [Custom Alphabet Parameters] \emph{Custom Alphabet} characters
are \\those that employ a user-specified alphabet. With this data
type, only insertion-deletion and substitution events are allowed.
\emph{Custom Alphabet} characters can be input as prealigned.  Within
this subsection, the user must specify the heuristic \emph{Level}
of the median sequence calculation. \emph{Direct Optimization} is
employed in median sequence calculation. Because calculating the
median states between custom alphabet strings becomes more
computationally intensive (and time consuming) as the number of
elements in the alphabet increases, the user should select a heuristic
level of median calculation appropriate for their data.  The default
level is 2.

\indent In addition to the data file, the user is require to upload
a \emph {Cost Matrix} that specifies the substitution and indel
transformation costs for alphabet elements.  By selecting the \emph
{Cost Matrix} button within this subsection, the user can upload a
cost matrix that specifies these costs for their custom alphabet
data. For details on the format requirements for custom alphabet
data files and their associated cost matrices see the
argument~\nccross{custom\_alphabet}{customalphabet} within the
command  \poycommand{read}.

\item [Break Inversion Parameters] \emph{Break Inversion} characters
are an enhancement of \emph{Custom Alphabet} characters. In addition
to allowing substitution and insertion deletion events, element
rearrangements, as well as orientation information can also be
optimized.  The median solvers provided restrict the analysis of
prealigned data.  The rearrangement costs for \emph{Break Inversion}
characters can be optimized using either \emph{Breakpoint} or
\emph{Locus Inversion} approaches (see the
arguments~\nccross{locus\_breakpoint}{chromosomelocusbreakpoint}
and~\nccross{locus\_inversion} {chromosomelocusinversion} within
the \emph{Chromosome and genome transformation methods} of the
command \poycommand{transform}). The default cost for both is set
to 10. A \emph{Median solver} must also be specified for the
optimization of rearrangements. The default median solver is
\emph{Caprara} \cite{Caprara2001}, but the user can alternatively
choose \emph{BBTSP, ChainedLK, COALESTSP, MGR}
\cite{bourqueandpevzner2002}, \emph{Siepel} \cite{siepelmoret2001},
\emph{SimpleLK} and \emph{Vinh} (the TSP solvers BBTSP, CoalesTSP,
ChainedLK and SimpleLK taken from the Concorde package)
(see~\nccross{median\_solver} {chromosomemediansolver}  within the
command \poycommand{transform}).  The calculation of median states
between \emph {Break Inversion} strings becomes more computationally
intensive (and time consuming) as the number of elements in the
alphabet increases, therefore a single heuristic level of median
calculation can only be employed for these character types--the
default level is 1.

\indent The requirements for \emph{Break Inversion} character types
are identical to those for \emph{Custom Alphabet} characters, with
respect to substitution and indel transformation costs.  By selecting
the  \emph{Cost Matrix} button within this subsection, the user can
upload a cost matrix to specify these costs. The user should see
the argument~\nccross{custom\_alphabet} {customalphabet} within the
command \poycommand{read} for details on the format requirements
for these cost matrices, which are identical in form to those for
\emph{Custom Alphabet} characters.

\item [Qualitative Parameters] \emph{Qualitative} data are any
non-sequence, prealigned data type (e.g. morphology, behavior).
These character types are optimized as additive, non-additive or
Sankoff characters and this information must be included in the
data file when using the \emph{Graphical User Interface}.

\end{description}

\hangindent=1cm  \emph{Likelihood Optimality Criterion} Currently,
it is only possible to perform a likelihood analysis of DNA sequences
(prealigned being permitted), %morphology and qualitative character
types. Using the \emph{GUI}, these data can be analyzed with
\texttt{likelihood} only (currently, transformation of the data to
\texttt{elikelihood} cannot be performed using the \emph{GUI}).
More detailed explanations of these options can be found below and
in the \emph{Likelihood transformation methods} section of
~\ccross{transform}.  In this section the user can specify the
\emph{likelihood model} of character substitution under which the
analysis will be performed.  Available substitution models include
JC69/Neyman, F81, K2P/K80, F84, HKY85, TN93, GTR, and NCM.  Users
can also perform phylogenetic model selection using AIC, AICc and
BIC.  Within this section it is possible to specify the nature of
among-site variation under \emph{Rate Distribution}.  Rate variation
distributions allow multipliers to be applied to separate groups
of characters.  These distributions can be set to \emph{Constant},
\emph{Gamma} (for non-zero rate variation) or  \emph{Theta} (for
parameterization of invariant sites).  These distribution values
can be specified for all of the available models. In addition,
\emph{Rate Classes} enables the user to specify the number of rate
classes for the discrete \emph{Gamma} rate distribution.  The user
can choose up to 8 rate classes for either \emph{Gamma} or \emph{Theta}
models (the default is 4).  \emph{Gap Treatment}  specifies the
treatment of indels.  There are three options \emph{missing},
\emph{character}, and \emph{character plus coupled}.  When gaps are
treated as \emph{missing}, they play no role in the calculation of
tree likelihoods.  The \emph{character} option treats the insertion
and deletion of A, C, G, and T each as different types of events
that are independently estimated (hence additional parameters over
\emph{coupled}). When \emph{coupled} is specified, all indel events
are treated with the same rate parameter.  \emph{Cost} specifies
the form of likelihood employed.  The options for prealigned data
(sequence and qualitative) are \emph{MAL} for \emph{maximum average
likelihood} and {MPL} for \emph{most parsimonious likelihood}
\cite{barryandhartigan1987}. Unaligned sequences can be optimized
with both \emph{MAL} (by transforming the sequences to
\poyargument{fixed\_states} first) and \emph{MPL}.  The prior
probabilities of the states are determined using \emph{Priors}.
The options are \emph{equal}, where each state prior is set to 1
divided by the number of states, and \emph{estimated}, where the
priors are set by their frequency in the data set.  Priors for gaps
are estimated by the maximum difference in length between input
sequences.

\begin{description}
\setlength{\parindent}{0.5cm}	   
\item[Output Files]
Designates the names and locations of files containing results of
the analysis.  By default, all of these output options are generated
with default names applied.  The names can then be changed in the
generated script or the option can be removed entirely.  As implied
by their respective titles, the \emph{tree} buttons output trees
in both parenthetical (best and consensus trees) and postscript
form (although this button only outputs a PDF file of the optimal
trees found, a useful commands that the user can include in the
generated script is to output a PDF file of the consensus tree (see
the argument~\nccross{graphconsensus}{graphconsensus} within the
command \poycommand{report})).

\indent A \emph{diagnosis} file provides information relating to
the analysis. Information in this file includes the cost of the
tree, the rearrangement costs (in the case of the analysis of
chromosome, genome and break inversion data types), as well as
information about each resulting node in the tree.  At each node,
the user is provided with a cost of the tree down to that node, a
rearrangement cost (if applicable), the ``descendant nodes'' coming
from this node and information concerning individual characters at
these nodes.

\begin{statement} The \texttt{root} in the diagnosis file may not
be the same as the root \poyargument{set} by the user.  This is
because the tree length heuristics may be based on an alternate
rooting scheme than that used for the \poyargument{newick} or
\poyargument {graphic trees} output.  \end{statement}

\indent The \emph{Analyzed data information} outputs a summary of
the input data. Specifically, the number of terminals to be analyzed,
a list of included terminals with numerical identifications, list
of synonyms (if specified), a list of excluded terminals, the number
of included characters in each character-type category (additive,
non-additive, Sankoff and sequence) with the corresponding cost
regimes, a list of excluded characters, and a list of input files.

\indent The \emph{Outgroup} field allows the the user to specify
the outgroup taxon.  The name of the taxon should reflect the name
as interpreted by \poy.  Therefore, the name should take into account
synonymy files, and taxon names that contain commented out information
via use of a \$ sign (see the argument~\nccross{rename}{rename}).
\end{description}

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/simplesearch_window_filled.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/simplesearch_script.jpg}
\end{minipage}

\caption{The \emph{Simple Search} window with specified search 
parameters (left) and the corresponding \emph{Script Editor} window. 
Observe that the names of the output files are left as the default output names.}
\label{fig:ScriptEditor_Window}
\end{figure}

Once all the parameters are selected, click the \emph{Make Script} button 
and another window--the \emph{Script Editor}--containing the generated 
script, appears on screen (Figure~\ref{fig:ScriptEditor_Window}). 
The script can be edited by typing in the commands directly in the 
\emph{Script Editor} window, saved (by clicking the \emph{Save As} 
button), or replaced with another script (using the \emph{Open} button). 
To start the analysis, click the \emph{Run} button in the \emph{Script 
Editor} window. When the \emph{Run} button is clicked, \poy will issue a
request to save the script. Thus, not only does \poy execute the script but
it also creates the record of the type of analysis (including all user-defined 
specifications) that was performed. Moreover, these scripts can later be 
executed manually in the \emph{ncurses} or \emph{flat} interfaces, or
selected as a script to run in the \emph{Graphical User Interface}.

\subsubsection*{\emph{Timed Search}}

{\emph{Timed Search} (Figure~\ref{fig:timed_search}) implements a
default search strategy that effectively combines tree building
with TBR branch swapping, parsimony ratchet, and tree fusing.  The
\emph{Timed Search} window has the same four parameter groups
described for the \emph{Simple Search}. However, the \emph{Search
Parameters} section (called \emph{Search and Perturb Parameters})
contains four fields specifying the search targets instead of the
\emph{Repetitions} field. These include the following:


\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/timedsearch_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/timedsearch_window.jpg}
\end{minipage}

\caption{The \emph{Timed Search} window. Selecting \emph{Timed Search} 
from the \emph{Analysis} menu (left) and viewing the \emph{Timed Search} 
window options (right).}
\label{fig:timed_search}
\end{figure}

\begin{description}
\item[Maximum time] The maximum total execution time for the search. 
The time is specified as days:hours:minutes.
\item[Minimum time] The minimum total execution time for the search. 
The time is specified as days:hours:minutes.
\item[Maximum memory] The maximum amount of memory allocated 
for the search.
\item[Minimum hits] The minimum number of times that the minimum 
cost must be reached before terminating the search.
\end{description}

This heuristic search is a powerful tool for analyzing data. The
number of rounds of successive searching is limited only by the
previously specified search targets. Therefore, when performing a
\emph{Timed Search}, it is {\bf crucial} to set the maximum time
such that the program has a reasonable amount of time to perform a
search.  Thus, it is important to have some approximation as to the
length of time it would take to perform a single round of searching
(e.g. build (1), followed by TBR, ratchet and fusing in the case
of a parsimony analysis of DNA sequence data).  Clearly, this is
data and optimality criterion dependent.  With this information,
the user can then estimate the amount of time necessary to perform
a thorough search (perhaps 10 times the amount of time it took to
perform this single round of build, swap, ratchet and fusing).  The
user should also allow some time for the program to collate and
write the results to files.  If the user has opted to run this
analysis in parallel, this can take some time.

\subsubsection*{\emph{Search with Ratchet}}

The parsimony ratchet is a heuristic strategy to escape  local
optima during tree searching~\cite{Nixon1999}. The ratchet reweights
a given percentage of characters for a specified number of iterations
of a search. An analysis is then performed and the resulting tree
topology is evaluated using the original data matrix with all
characters (with original weights) to determine the length of the
tree. The \emph{Search with Ratchet}
(Figure~\ref{fig:search_with_ratchet_window}) follows the same basic
steps of a simple search but includes the ratchet step after the
swap. In addition to the same sequence alignment and search parameters
as described for the \emph{Simple Search} window, the \emph{Search
Parameters} section provides the following ratchet parameters fields:


\begin{description}
\item[Ratchet iterations] The number of iterations for the parsimony
ratchet.
\item[Severity] The severity parameter of the ratchet (the weight
change factor for the selected characters).
\item[Percentage] The percentage of characters to be reweighted during 
ratcheting.
\end{description}

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/searchwithratchet_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/searchwithratchet_window.jpg}
\end{minipage}

\caption{The \emph{Search with Ratchet} window. Selecting \emph{Search 
with Ratchet} from the \emph{Analysis}  menu (left) and viewing the 
\emph{Search with Ratchet} window options (right).}
\label{fig:search_with_ratchet_window}
\end{figure}

\subsubsection*{\emph{Search with Perturb}}

\emph{Search with Perturb} (Figure~\ref{fig:search_with_perturb_window})
provides an alternative means to escape local optima by changing
the transformation cost matrix of the sequence characters, a procedure
similar in spirit to the parsimony ratchet. In addition to the same
sequence optimization and search parameters as described for the
\emph{Simple Search} window, the \emph{Search with Perturb} window
provides three extra fields with the parameters for the transformation
cost matrix perturbation as follows:

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/searchwithperturb_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/searchwithperturb_window.jpg}
\end{minipage} 
\caption{The \emph{Search with Perturb} window. Selecting \emph{Search with 
Perturb} from the \emph{Analysis} menu (left) and viewing the \emph{Search 
with Perturb} window options (right).}
\label{fig:search_with_perturb_window}
\end{figure}

\begin{description}
%\setlength{\labelwidth}{0pt}
%\setlength{\labelsep}{5pt}
%\setlength{\itemindent}{0pt}%
\item[Perturb iterations] Sets the number of perturb iterations to be performed.
\item[Substitutions] Specifies the cost of the perturbed substitutions.
\item[Indels] Specifies the cost of the perturbed indels.
\end{description}

During this heuristic search, \poy performs a parsimony ratchet search during 
each iteration (default values are used, i.e. 25\% probability, 2 severity).

\subsubsection{Support calculation options}

It is possible to calculate several support values using this
interface.  Two of these measures, Bootstrap and Jackknife, involve
resampling techniques, while the third, Bremer support, is an
optimality-based measure based on the cost of the tree.

Although it is possible to calculate Jackknife and Bootstrap support
values for trees constructed using dynamic homology characters, it
is recommended against doing so as resampling of dynamic characters
occurs at the fragment, rather than nucleotide, level. Consequently,
the bootstrap and jackknife support values calculated for dynamic
characters are not directly comparable to those calculated based
on static character matrices. If it is desired to perform character
sampling at the level of individual nucleotides, the dynamic
characters must be transformed into static characters using
\poyargument{static\_approx} argument of the command~\ccross{transform}
prior to executing \poycommand{calculate\_support}.  Of course, if
the dataset of dynamic characters contains a large number of
fragments, this caveat may not be warranted.

For chromosome and genome character types, only the calculation of
Bremer support values is recommended.

None of the support calculation windows include functions for tree
building and searching. Therefore, one of the input files must
contain trees for which support values are going to be calculated.

\subsubsection*{\emph{Bootstrap}}

As a resampling technique, the non-parametric \emph{Bootstrap}
resamples the original data (with replacement), creating a simulated
dataset equal to the size of the original dataset. The \emph{Bootstrap}
window (Figure~\ref{fig:bootstrap}) specifies parameters for
estimating the Bootstrap support values. In addition to the
\emph{Simple Search} window fields, it contains a field for the
bootstrap parameters, in this case a \emph{Pseudoreplicates} field,
to specify the number of bootstrap pseudoreplicates.

\begin{description}
\item[Pseudoreplicates] Specifies the number of resampling iterations.
\end{description}

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/bootstrap_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/bootstrap_window.jpg}
\end{minipage}
\caption{The \emph{Bootstrap} window. Selecting \emph{Bootstrap} 
from the \emph{Analysis} menu (left) and viewing the \emph{Bootstrap} 
window options (right).}
\label{fig:bootstrap}
\end{figure}

\subsubsection*{\emph{Jackknife}}

An alternative statistical measure of support is the \emph{Jackknife}, 
wherein the original data matrix is resampled, but in this case without 
replacement.  The \emph{Jackknife} window (Figure~\ref{fig:jackknife}) 
specifies parameters for estimating the Jackknife support values. In 
addition to the \emph{Simple Search} window fields, \emph{Jackknife 
Parameters} contains fields to specify the number of Jackknife 
pseudoreplicates (\emph{Pseudoreplicates}) and the number of characters 
to be removed (\emph{Remove}) during each pseudoreplicate.

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/jackknife_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/jackknife_window.jpg}
\end{minipage}
\caption{The \emph{Jackknife} window. Selecting \emph{Jackknife} 
from the \emph{Analysis} menu (left) and 
viewing the \emph{Jackknife} window options (right).}
\label{fig:jackknife}
\end{figure}

\begin{description}
\item[Pseudoreplicates] Specifies the number of resampling iterations.
\item[Remove] Specifies the percentage of characters being deleted 
during a pseudoreplicate.
\end{description}

\subsubsection*{\emph{Bremer}}

As an optimality-based measure of calculating tree support, Bremer
values are the number of extra steps required before a clade is
lost in the most parsimonious or strict consensus of the most
parsimonious trees.  Bremer support under likelihood is equivalent
to the log of the likelihood ratios for each branch~\cite{Wheeler2006}.
There are two ways to determine Bremer support values in \poy.  The
first involves performing a series of searches, where each group
supported on the examined cladogram is constrained not to occur in
the result.  The second does not involve constraining the tree but
collects information for all the clades not present in the set
visited trees.  Currently, calculating support via ``visited'' trees
can only be performed sequentially and not parallel.

The \emph{Bremer} option (Figure~\ref{fig:search_for_bremer_menu})
is divided into two windows: the \emph{Search for Bremer} window,
that specifies the Bremer support \cite{Bremer1988, Kallersjoetal1992}
calculation parameters, and the \emph{Report Bremer} window to
format the output of the results (Figure~\ref{fig:search_report_bremer}).

\paragraph{Search for Bremer}

\begin{figure}[htpb]
\begin{center}
\includegraphics[width=0.65\textwidth]{doc/figures/searchforbremer_menu.jpg}
\end{center}
\caption{ Selecting the \emph{Bremer} windows from the \emph{Analysis} menu.}
\label{fig:search_for_bremer_menu}
\end{figure}

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/searchforbremer_window.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/reportbremer_window.jpg}
\end{minipage}
\caption{Viewing the options of the \emph{Search for Bremer} (left) and the 
\emph{Report Bremer}(right) windows.}
\label{fig:search_report_bremer}
\end{figure}

The script produced in this window collects trees visited during a search for 
Bremer support calculations. This search can take a long time, as the goal of 
this search strategy is to broadly sample variation among trees, and guarantee 
that all clades have Bremer support values.  

In addition to the standard four sections defined for the \emph{Simple Search} 
window, that one of the output files is the \emph{Temporary Trees} file, which 
contains all the information required to produce the Bremer support tree
results in the \emph{Report Bremer} window. Make sure to choose a file 
name that does not overwrite this output.

If the search does not finish within the time frame available to the user the search 
can be interrupted and the intermediate results remain stored in the \emph{Temporary 
Trees} file.  As Bremer calculations are upper-bound values, terminating the search 
prior to completion and, thus, storing a smaller pool of visited trees may inflate 
support values relative to those generated by a more exhaustive search. The trees 
from the \emph{Temporary Trees} file can then be reported using the 
\emph{Report Bremer} window.

\paragraph{Report Bremer}
The script produced in this window takes the \emph{Temporary Trees} file 
generated in the \emph{Search for Bremer} window in the \emph{File with 
trees for Bremer calculation} field. 

\subsubsection{Diagnosis}
\subsubsection*{\emph{Diagnose Tree}}

The \emph{Diagnose Tree} window (Figure~\ref{fig:diagnosetree}) specifies 
parameters for reporting a diagnosis of the input tree. This window lacks the 
\emph{Search Parameters} section because the diagnosis is performed on 
the trees resulted from prior searches and no new trees are generated during 
the diagnosis procedure. In addition to the tree (or trees) file, the user must have 
input the data file associated with this tree file, in order to diagnose
the tree(s) in this file.

\begin{figure}[ht]
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/diagnose_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/diagnose_window.jpg}
\end{minipage}
\caption{The \emph{Diagnose} window. Selecting \emph{Diagnose Tree} 
from the \emph{Analysis} menu (left) and viewing the \emph{Diagnose} 
window options (right).}
\label{fig:diagnosetree}
\end{figure}

\subsubsection{\emph{Script editing and the Interactive Console}}

\subsubsection*{Open POY Launcher}

Selecting \emph{Open POY Script} (Figure~\ref{fig:open_poy_launcher}) 
displays the \emph{POY Launcher} window (Figure~\ref{fig:menu_launcher_window}), 
the function of which is described above.

\begin{figure}[htpb]
\begin{center}
\includegraphics[width=0.5\textwidth]{doc/figures/openpoylauncher_menu.jpg}
\end{center}
\caption{The \emph{Open POY Launcher} selection opens the \emph{POY 
Launcher} window.}
\label{fig:open_poy_launcher}
\end{figure}

\subsubsection*{Run Interactive Console}

Selecting \emph{Run Interactive Console} (Figure~\ref{fig:runinteractive}) 
opens the \emph{ncurses} interface that enables the user to run the analysis 
interactively by entering \poy commands directly via the command-line 
interface of the \emph{Interactive Console} See \emph{Using the 
Interactive Console} (Section \ref{interactiveconsole}).

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/runinteractive_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/create_script_menu.jpg}
\end{minipage}
\caption{The \emph{Run Interactive Console} selection (left) opens \poy 
interactive console in a new window. The \emph{Create Script} selection 
opens the \emph{Script Editor} window (Figure~\ref{fig:ScriptEditor_Window}).}
\label{fig:runinteractive}
\end{figure}

\subsubsection*{Create Script}
The \emph{Create Script} selection opens a blank \emph{Script Editor} 
window that allows opening, creating, modifying, saving, and executing  
a customized script.

\subsection{The \emph{View} menu}

The \emph{View} menu contains the \emph{Output} window which is 
subdivided into two fields: the upper \emph{Results and Errors} and lower 
\emph{Status of Search} (Figure~\ref{fig:results_and_status_windows}). 
These fields display, respectively, the results (including warning and error 
messages) and the current state of the analysis. These fields are not updated 
automatically and in order to display the current state of the analysis the 
user must click the \emph{Update} button. The \emph{View} menu also 
contains the \emph{About POY} window in Windows.

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/view_menu.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/output_window.jpg}
\end{minipage}
\caption{Selecting the \emph{Output} window (left) and viewing the 
\emph{Results and Errors} and  \emph{Status of Search} fields.}
\label{fig:results_and_status_windows}
\end{figure}

%-----------------------------------
%Interactive Console
%-----------------------------------
\section{Using the Interactive Console} \label{interactiveconsole}

This section will help you get started using the \poy \emph{Interactive 
Console} and will prepare you for the more extensive, technical descriptions 
in the next chapter, \emph{\poy Commands}.
This section will illustrate how to input data files, check the data you are 
analyzing, generate a set of initial trees, do basic branch 
swapping to find a local optimum, and, finally, produce and visualize 
the resultant trees, their consensus, and generate support 
values in a command-line environment rather than using a 
\emph{Graphical User Interface}. 

For the purpose of this exercise, two data files, which are included in the 
download package, are available at the \poy download page.\\
\begin{center}
\url{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy/download}
\end{center}

\begin{itemize}
\item {\texttt{28s.fas} contains unaligned DNA sequences (partial 28S 
ribosomal RNA) in FASTA format.~\cite{pearson1988}}
\item {\texttt{morpho.ss} contains a morphological data matrix in 
Hennig86 format.~\cite{farris1988}}
\end{itemize}

Once \poy has been launched and the interface (Figure~\ref{fig:figinterface})  
appears on the screen, data can be input and analysis can proceed. As you 
follow the instructions, you are encouraged to consult the help file by using 
the command \commandstyle{help} (see Section~\ref{sec:help} to learn 
more about \poy commands and their arguments).

\subsection{The interface}

The \emph{Interactive Console} provides a terminal environment with
enhanced ability to display the results and the state of the analysis.
We recommend the use of the console to explore and verify the data
in the early steps of the analysis, and to learn the scripting
language. Using the console requires familiarity with \poy commands,
their arguments, and the conventions of \poy scripting (which are
discussed in the \emph{POY Commands} chapter). It has four windows:
\emph{POY Output}, \emph{Interactive Console}, \emph{State of Stored
Search}, and \emph{Current Job} (Figure \ref{fig:figinterface}):


\begin{figure}[htbp]
\centering
\includegraphics[width=0.9\textwidth]{doc/figures/figinterface.jpg}
\caption{\poy interface displayed in the Terminal window prior to analysis. 
Observe the cursor at the \poy prompt in the \emph{Interactive Console} 
and note that the \emph{State of Stored Search} and \emph{Current Job} 
windows are empty.}
\label{fig:figinterface}
\end{figure}

\begin{description}
%\setlength{\labelwidth}{0pt}
\setlength{\labelsep}{5pt}
\setlength{\itemindent}{0pt}%
\item[POY Output] (Figure \ref{fig:figinterface}, upper box) displays the 
status of the imported data, outputs the results of the phylogenetic analyses 
(such as trees, character diagnoses, and implied alignments), reports errors, 
and displays descriptions of \poy commands.
\item[Interactive Console] (Figure \ref{fig:figinterface}, mid-left box) 
is used to issue the commands interactively and to execute the commands 
by clicking the Return key. (See Section~\ref{commands} on the description 
of \poy commands.)
\item[State of Stored Search] (Figure \ref{fig:figinterface}, mid-right box) 
displays the time (in seconds) elapsed since the initiation of the current 
operation. This window also reports the number of trees currently in m
emory and displays the range of their costs.
\item[Current Job] (Figure \ref{fig:figinterface}, lower box) describes the 
currently running operation. When the operation is completed, the box is blank.
\end{description} 

\begin{figure}[htbp]
\centering
\includegraphics[width=0.9\textwidth]{doc/figures/figprocess.jpg}
\caption{\poy \emph{Interactive Console} during a process. The \emph{POY Output} 
window displays (by default) the information on the input data files. The 
\emph{Interactive Console} lists the commands that have been executed. The 
\emph{Current Job} window shows the state of the current operation and the 
current tree score. The \emph{State of Stored Search} shows the time elapsed  
since the last command \commandstyle{swap}, was initiated.}
\label{fig:figprocess}
\end{figure}

\subsection{Starting a \poy session using the \emph{Interactive Console}}

\begin{flushleft}
\begin{minipage}[c]{0.074\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figlogowindows.jpg}
\end{minipage}
\,
\begin{minipage}[t]{0.88\textwidth}
\subsubsection*{Windows}
\end{minipage}
\begin{itemize}
\item{Start$>$All Programs$>$POY$>$Interactive Console}
\end{itemize}

\begin{minipage}[c]{0.074\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figlogomac.jpg}
\end{minipage}
\,
\begin{minipage}[t]{0.88\textwidth}
\subsubsection*{Mac OSX}
\end{minipage}
\begin{itemize}
\item {Double-click \poy application icon to start the program.}
\item {Select \emph{Run Interactive Console} from the
\emph{Analyses} menu.}
\end{itemize}		

\begin{minipage}[c]{0.074\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figlogolinux.jpg}
\end{minipage}
\,
\begin{minipage}[t]{0.88\textwidth}
\subsubsection*{Linux}
\end{minipage}
\begin{itemize}
\item Add \texttt{/opt/poy5/Resources/} (or the location you plan to install) to your \texttt{PATH} and run
\texttt{ncurses\_poy} from a terminal.
\end{itemize}
\end{flushleft}

\subsection{Entering commands}
Once this \poy interface is opened, the cursor appears in the
\emph{Interactive Console} portion of the window and is ready to
accept commands. The \emph{Interactive Console} does not support
using the mouse and, as is true for most command-line applications,
the cursor can be moved using the left and right arrow keys, and
the Backspace (in Windows) or Delete (in Mac) keys are used to erase
individual characters to the left of the current cursor position.
To eliminate the need for retyping commands anew during a \poy
session, keyboard shortcuts can be used: control-p (``previous'')
and control-n (``next'') will scroll through the commands previously
entered during the session. In addition, the \emph{Interactive
Console} is equipped with the autocomplete feature: it involves
\poy predicting a command, an argument, of file name that the user
wants to type from the first letter(s) entered. Upon typing the
first letter or part of the phrase, repeatedly pressing the TAB key
scrolls through the list of command, argument, and file names that
begin with that letter or phrase. Autocomplete simplifies interaction
with the program.

\subsection{Browsing the output}
As output is reported in the \emph{POY Output} window, only the
most recent reports will be seen in the window.  Using the Up and
Down keys allows the user to scroll up and down the \emph{POY Output}
window to see the welcome line, and previously printed reports and
help descriptions. Pressing Up and Down keys automatically places
the cursor in the lower left corner of the \emph{POY Output} window
indicating that you are interacting with that window. Only 1000
lines are stored in the memory and the output that was reported
before that will not be accessible by scrolling. The number of
lines, however, can be modified by the user using the command
\poycommand{set()}, see~\ccross{history}. If the user desires to
keep the entire output or specific items in the output, a log can
be created using the command \poycommand{set()}, see~\ccross{log})
or specific outputs can be redirected to files (see~\ccross{report}).
The user should be aware that outputting a log file can slow down
the program due to IO (input/output) delay.

\subsection{Switching between the windows}
To return to the \emph{Interactive Console}, start typing and the
cursor will automatically be placed back at the \poy prompt.  When
an operation is in progress (shown in the \emph{Current Job} window),
the cursor stays in the upper left corner of the \emph{State of
Current Search} window, and switching between the \emph{Interactive
Console} and the \emph{POY Output} window is disabled. There are
no user interactions in the \emph{Current Job} or \emph{State of
the State of Current Search}.

\subsection{Input of data} \label{sec:import}

The basic command to input data in \poy is \commandstyle{read()},
which includes the list of files (in quotation marks and separated
by commas) enclosed in parentheses. Suppose that we would like to
simultaneously analyze morphological and molecular datasets, contained
in separate data files, \texttt{morpho.ss} and \texttt{28s.fas},
respectively. We can issue a pair of \commandstyle{read()} commands
(Figure~\ref{fig:readingexample}): 

\begin{quote}
\commandstyle{read("morpho.ss")}\\ 
\commandstyle{read("28s.fas")}
\end{quote}

\begin{figure}
\begin{center}
\includegraphics[width=0.9\textwidth]{doc/figures/reading_example.jpg}
\end{center}

\caption{Importing data files using the \emph{Interactive Console}.
Two consecutive \commandstyle{read} commands specify both the
morphological data file in Hennig86 format (\texttt{morpho.ss}),
and the molecular data file in FASTA format (\texttt{28s.fas}).
Observe that \poy automatically reports  in the \emph{POY Output}
window the names and types of files that have been imported.}
\label{fig:readingexample} \end{figure}

The syntax of \commandstyle{read}, like every command in \poy,
contains two elements: the name of the command, in this case
\commandstyle{read}, followed by an optional list of arguments
separated by commas and enclosed in parentheses. All filenames read
into \poy should include the appropriate suffix for the file type
(e.g. .fas, .ss, .aln, .tre etc:).  Typically, the arguments of the
command \commandstyle{read()} are names of data files, each being
enclosed in double quotes (as shown in the example above). Even
though there might be only one argument or none in some commands,
parentheses (e.g. \poycommand{pwd()}) always follow the command
name. An exhaustive discussion of \poy command structure and detailed
descriptions of all commands with examples of their usage are
provided in the \emph{POY Commands} chapter (\ref{commands}).

In order to import data by entering the names of the files, the
directory containing these files must be identified.  This can be
established in two ways--by using the command \commandstyle {cd}
to redirect the path to the directory where the data are found and
then reading in the data file:\\ 

\begin{quote}
\commandstyle{cd("/Users/username/docs/poyfiles")}\\
\commandstyle{read("28s.fas")}\\ 
\end{quote} 

or by including the
full path in the argument of \commandstyle{read}:\\ 

\begin{quote}
\commandstyle{read("/Users/username/docs/poyfiles/28s.fas")}
\end{quote}

Most of the time users are interested in importing multiple data
files to analyze an entire dataset. In this case, multiple data
files can be specified as arguments for a single command. For
example, importing both files, \texttt{morpho.ss} and \texttt{28s.fas},
can be written more succinctly:\\ \begin{quote}
\commandstyle{read("morpho.ss","28s.fas")} \end{quote} or if the
full path is included in the argument of \commandstyle{read} as:
\\

\begin{quote}
\commandstyle{read("/Users/username/docs/poyfiles/morpho.ss", \\
"/Users/username/docs/poyfiles/28s.fas")}
\end{quote}

This is equivalent to sequentially importing each file as shown in
(Figures ~\ref{fig:readingexample} and \ref{fig:reading_example2}).

Figures~\ref{fig:readingexample} and \ref{fig:reading_example2}
also illustrate an important feature that makes \poy different from
many other phylogenetic analysis programs: every time a file is
imported during a \poy session, the input data are \emph{added} to
the data in memory and \emph{do not replace them}. This
allows additional analytical flexibility. For example, if only
morphological data are read and trees are built based on these data
alone, a subsequently imported molecular character dataset will be
used in conjunction with the previously imported morphological data,
despite the fact that current trees in memory were generated only
from morphological data (Figure~\ref{fig:reading_example2}):

\begin{quote}
\commandstyle{read("morpho.ss")}\\
\commandstyle{build()}\\
\commandstyle{read("28s.fas")}\\
\commandstyle{rediagnose()}\\
\commandstyle{swap()}
\end{quote}

\begin{figure}[]
\begin{center}
\includegraphics[width=0.9\textwidth]{doc/figures/reading_example2.jpg}
\end{center}
\caption{Building trees with morphological data only but continuing
the analysis using combined morphological and molecular data. This
example shows how we can add data to the analysis incrementally by
loading files at different points in the search. First, the
morphological data are imported from \texttt{morpho.ss} file using
\poycommand{read()} and trees are built based on these data.
Then molecular data from the \texttt{28s.fas} file are loaded into
memory. Finally, subsequent analyses, \commandstyle{rediagnose()} and
\commandstyle{swap()}, are conducted using all the data in memory, that
is the trees based on morphological data, and both morphological
and molecular character sets.} \label{fig:reading_example2}
\end{figure}

It must be noted that if the numbers of terminals differ among data
files, \emph{only} the data that correspond to the terminals used
to generate the trees (in this case, the morphological data file)
are used. The rest of the character data are ignored, unless the
trees are built again with the data files containing the expanded
number of terminals.  Also, because \poy appends trees and data in
memory, it is a good practice when starting a new analysis within
the same interactive session to clear the data using the command 
\commandstyle{wipe()}.

Valid input files include nucleotide and amino acid sequence files
in many formats, and morphological data in Hennig86 and Nexus
formats. (For information on specific formats supported by \poy and
other types of input files see ~\ccross{read}.)

\subsection{Inspecting data}

Once a dataset (or multiple datasets) is imported, \poy automatically
reports a brief description of contents for each loaded file in the
\emph{POY Output} (Figure ~\ref{fig:readingexample}). However, it
may be desirable to inspect the imported data in greater detail to
ensure that the format and contents of the files have been interpreted
correctly. This practice helps avoid common errors, such as
inconsistently spelled terminal names, which may result in bogus
results, produce error messages, and aborted jobs.

The basic command for outputting information is \commandstyle{report()}.
One of its arguments, \commandstyle{data}, outputs a set of tables
showing the list of terminals, the number and types of characters,
and the lists of terminals and characters excluded from the analysis.
To produce a report of the data files that were used in the previous
example (\texttt{morpho.ss} and \texttt{28s.fas}), we import the
data and execute \commandstyle{report(data)}:

\begin{quote}
\commandstyle{read("morpho.ss","28s.fas")}\\
\commandstyle{report(data)}
\end{quote}

This will generate an extensive, detailed output, partial views of
which are shown in Figure ~\ref{fig:reportdata}.  Obviously, the
entire report will not be visible in the \emph{POY Output} window.
Therefore, the Up and Down arrow keys and Page Up and Page Down
keys can be used to scroll.  By default, \poy reports the results
of executed commands to the \emph{POY Output} window. However, the
same output can be redirected to a file simply by adding the name
of the output file in the list of argument of the command
\commandstyle{report()} \emph{before} the argument specifying the
type of the requested report (in this case \commandstyle{data}, see
the command~\ccross{report} ). For instance, to output the data
into the file \texttt{data\_analyzed.txt} we would enter:

\begin{quote}
\commandstyle{read("morpho.ss","28s.fas")}\\
\commandstyle{report("data\_analyzed.txt",data)}
\end{quote}

\begin{figure}
\centering
\begin{minipage}[c]{0.52\textwidth}
\includegraphics[width=\textwidth]{doc/figures/report2.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.44\textwidth}
\includegraphics[width=\textwidth]{doc/figures/report3.jpg}
\end{minipage}
\caption{Inspecting imported data. The figure shows segments of 
a data report generated by \commandstyle{report(data)}. 
The left and right panels demonstrate a typical table output 
the character and terminal data respectively.}
\label{fig:reportdata}
\end{figure}

In this example, all the imported data are analyzed and, therefore,
the report fields that list excluded data will appear empty.  One
can, however, exclude specific characters or terminals from the
analysis using additional commands (see the command~\ccross{report}).

Another useful argument of \commandstyle{report} is
\commandstyle{cross\_references}. This argument displays whether
character data are present or absent for each terminal in each one
of the imported data files. This provides a comprehensive visual
overview of missing data. Building on the previous example, such
output can be generated by the following sequence of commands:
\begin{quote} \commandstyle{read("morpho.ss","28s.fas")}\\
\commandstyle{report("cross\_refs.txt",cross\_references)} \end{quote}

\begin{figure}[]
\begin{center}
\includegraphics[width=1.0\textwidth]{doc/figures/crossref.jpg}
\end{center}
\caption{Visualizing missing data. The command
\commandstyle{cross\_references} displays a table showing whether
a given terminal (in the left column) is present (``+'') or absent
(``--'') in each data file. In this example, \texttt{28s.fas} is
missing for Amblypygid and \texttt{morpho.ss} for Hypochilus.}
\label{fig:crossref} \end{figure}

A typical output of \commandstyle{cross\_references} command is
shown in Figure ~\ref{fig:crossref}. This argument is a very useful
tool for visual representation of missing data. Moreover, reporting
all the data to a cross references file can also highlight
inconsistencies in the spelling of taxon names in different data
files.

\subsection{Building the initial trees}

The command to build trees is \commandstyle{build()} (already
mentioned in Section~\ref{sec:import}). After importing \texttt{morpho.ss}
and \texttt{28s.fas}, executing the command \commandstyle{build()}
without specifying any arguments (default settings) generates 10
Wagner trees by random addition sequence.

Many \poy commands operate under default settings when executed
without arguments. To learn what the default settings are for a
particular command use either the \commandstyle{help()} command with
the command name of interest inserted in parentheses or consult the
\emph{POY Commands} chapter (\ref{commands}).

If the user would like to specify a number of tree building replicates
different from the default value of 10, the argument \commandstyle{trees}
followed by a colon (``:'') and an integer specifying the number
of trees must be included in the argument list of the \commandstyle{build}
command: \commandstyle{build(trees:100)}. This command has a shortcut
that omits the argument \commandstyle{trees}. Thus,
\commandstyle{build(trees:100)} is equivalent to \commandstyle{build(100)}.
As defaults, the shortcuts are fully described in Section \ref{commands}.
The entire sequence of commands minimally required to import the
data and build 100 trees is the following:

\begin{quote}
\commandstyle{read("morpho.ss","28s.fas")}\\
\commandstyle{build(100)}
\end{quote}

As the tree building advances, the \emph{Current Job} window displays
the current status of the operation (Figure~\ref{fig:building}).
This window shows how many Wagner builds have been performed out
of the total number requested, the number of terminals added in the
current build, the cost of the current tree (recalculated after
each terminal addition), and the estimated time for the completion
of all the builds. When all the trees are generated, the \emph{State
of Stored Search} window displays the range of tree costs (the best
and worst costs), the number of trees stored in memory, and the
number of trees with the best cost (Figure~\ref{fig:building}).

\begin{figure}
\centering
\begin{minipage}[c]{0.507\textwidth}
\includegraphics[width=\textwidth]{doc/figures/building1.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.453\textwidth}
\includegraphics[width=\textwidth]{doc/figures/building2.jpg}
\end{minipage}
\caption{Generating Wagner trees. During the process of tree building (left 
panel), the \emph{Current Job} window displays how many builds have
been performed so far (\texttt{57 of 100}), the number of terminals
added in the current build (\texttt{13 of 17}), the cost of a current
tree recalculated after each terminal addition (\texttt{362}), and 
the estimated time (in seconds) for the completion of the operation
(\texttt{4 s}). Because the process is not complete, the \emph{State
of Stored Search} window contains no trees. Once tree building is
complete, the \emph{State of Stored Search} window displays the
best (\texttt{451}) and worst (\texttt{472}) costs, the number of
trees stored in memory (\texttt{100}), and the number of trees with
the best cost (\texttt{2}).} 
\label{fig:building}
\end{figure}

\subsection{Performing a local search}

Now that the trees have been generated and stored in memory, a local
search can be performed to refine and improve the initial trees by
examining additional topologies of potentially better cost.  The
command \commandstyle{swap()} implements an efficient strategy by
performing SPR and TBR branch swapping alternately. As with other
commands, the arguments of \commandstyle{swap()} allow the customization
of the swap algorithm. In the following example, branch swapping
is performed under the default settings on each of the 100 trees
build in the previous step:

\begin{quote}
\commandstyle{read("morpho.ss","28s.fas")}\\
\commandstyle{build(100)}\\
\commandstyle{swap()}
\end{quote}

Branch swapping is performed sequentially on all trees stored in
memory. During swapping, the \emph{Current Job} window reports the
number of the tree that is currently being analyzed, the method of
branch swapping, the specific routine being currently performed,
and the cost of the current tree (Figure~\ref{fig:swapping}). When
the process is complete, the \emph{State of Stored Search} window
displays the range of tree costs (the best and worst costs), the
number of trees stored in memory, and the number of trees with the
best cost (Figure~\ref{fig:swapping}). Notice that the local search
had reduced the costs of the initial best (from 451 to 446) and
narrowed the range of tree costs.

\begin{figure}
\centering
\begin{minipage}[c]{0.49\textwidth}
\includegraphics[width=\textwidth]{doc/figures/swap1.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.453\textwidth}
\includegraphics[width=\textwidth]{doc/figures/swap2.jpg}
\end{minipage}
\caption{Performing a local search. When searching (left panel),
the \emph{Current Job} window reports the number of the tree that
is currently being analyzed (\texttt{73 of 100}), a method of branch
swapping (\texttt{Alternate}), a function being currently performed
(\texttt{SPR search}), and a cost of the current tree (\texttt{456}).
When the searching is finished (right panel), the \emph{State of
Stored Search} window displays the best (\texttt{446}) and worst
(\texttt{463}) costs, the number of trees stored in memory
(\texttt{100}), and the number of trees of the best cost (\texttt{9})
recovered from independent tree builds and swaps. Notice that these trees may
not necessarily have unique topologies.}
\label{fig:swapping}
\end{figure}

Using different combinations of \commandstyle{swap()} arguments
allow the designation of a  large number of search strategies with
different levels of complexity. Some simple options allow the choice
between SPR and TBR. More complex strategies allow keeping a specific
number of best trees per single initial tree (generated during the
building step).  For example, the command \commandstyle{swap(trees:10)}
will keep up to 10 best trees generated during branch swapping on
a single initial tree. Consequently, if 100 trees were built
initially, this command will produce up to 1,000 trees. The argument
\commandstyle{threshold} allows the retention of suboptimal trees
within a specified percent of cost difference from the current best
tree. For example, \commandstyle{swap(trees:20,threshold:10)} will
execute a swap considering trees within a ten percent cost difference
of the current best tree and retain the 20 minimal length swapped
trees for each initial build. Other options provide the means to
sample trees as they are evaluated, timeout after certain number
of seconds, transform the cost regime, and other functions in
conjunction with many \poy commands.

\subsection{Selecting trees}

Having performed the basic steps of importing character data,
building initial trees, and conducting a simple local search, we
obtained a set of locally optimal trees in memory. Generally, a user
would like to select only those trees that are both optimal \emph{and}
topologically unique. The default setting of the \commandstyle{select()}
does exactly that.  Adding \commandstyle{select()} to our example
of command sequence for the basic analysis

\begin{quote}
\commandstyle{read("morpho.ss","28s.fas")}\\
\commandstyle{build(100)}\\
\commandstyle{swap()}\\
\commandstyle{select()}
\end{quote}

selects only unique trees of best cost. The remaining trees are
deleted from memory. The \emph{State of Stored Search} window reports
the number and the cost of the best tree(s) (Figure~\ref{fig:select}).

As an alternative, the user may choose to select topologically
unique trees, regardless of the cost, using \texttt{select(unique)}.
This may ensure that a larger tree space is explored.  If this is
used as an option during the search, the user should remember to
\commandstyle{select()} at the end of the run, prior to reporting
the results.

\begin{figure}[]
\begin{center}
\includegraphics[width=0.9\textwidth]{doc/figures/select.jpg}
\end{center}
\caption{Selecting unique best trees. Executing \commandstyle{select()} 
keeps only unique trees of best cost. The \emph{State of Stored Search} 
window reports that there is only one unique tree of best cost (\texttt{446}).}
\label{fig:select}
\end{figure}

The command \commandstyle{select()}, is another multifunctional
command the arguments of which are also used to select (include or
exclude) specific terminals, characters, and trees.) Comparing the
output reported in the \emph{State of Stored Search} before
(Figure~\ref{fig:swapping}) and after (Figure~\ref{fig:select})
executing \commandstyle{select()} shows that swapping on 9 of 100
initial trees produced the trees of best cost (\texttt{446}), but
these trees are identical, because only one was retained when
filtered using \commandstyle{select()}.

\subsection{Visualizing the results}

There are several options for visualizing results in \poy
(see~\ccross{report}). The command \commandstyle{report("my\_first\_tree",graphtrees)} 
outputs a cladogram in PDF format (Figure~\ref{fig:trees}),
which can be displayed, edited, and printed using graphics software
(such as Adobe Illustrator or Preview).  \poy also appends the
``pdf'' extension when generating graphic output to a file. A quick
way to see the tree(s) on screen is to use the command
\commandstyle{report(asciitrees)} that draws a cladogram in the
\emph{POY Output} window (Figure~\ref{fig:trees}). The ascii tree(s)
can also be reported to a file, if an output file name is specified
within the command (\commandstyle{report("my\_first\_trees",
asciitrees)}).  These trees will be saved to a text file.

The command \commandstyle{report("my\_first\_trees.txt",trees)}
reports the trees in memory in parenthetical notation to the file
\texttt{my\_first\_trees} that can be imported in other programs.
Supported tree output formats include Newick and Hennig86.
\commandstyle{report()} can also generate consensus trees in the
graphical and parenthetical formats when appropriate arguments are
specified (for example, \commandstyle{report("strict\_consensus",graphconsensus)}).

\begin{figure}
\centering
\begin{minipage}[c]{0.45\textwidth}
\includegraphics[width=\textwidth]{doc/figures/asciitrees.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.5\textwidth}
\includegraphics[width=\textwidth]{doc/figures/pstree.jpg}
\end{minipage}
\caption{Visualizing trees. An ascii tree (left) is generated using the command
\poycommand{report(asciitrees)}. The same tree is reported to a file in a PDF 
format (right) using \commandstyle{report("my\_first\_tree",graphtrees)}. 
Observe that both representations of trees  are preceded by their costs.}
\label{fig:trees}
\end{figure}

\subsection{Interrupting a process}
To interrupt a process, press control-c. By default, an error,
\texttt{Error:}\\ \texttt{Interrupted}, is reported in the \emph{POY
Output} window. The program does not close, however, and a new
command can be entered. Interrupting the analysis cancels the
execution of the last command requested by the user and restores
the data and trees in memory before that last command. For example,
the following two session are equivalent:

\begin{quote}
\commandstyle{read("morpho.ss") <ENTER>} \\
\end{quote}
and

\begin{quote}
\commandstyle{read("morpho.ss") <ENTER>} \\
\commandstyle{read("28s.fas") <CONTROL-C>} \\
\end{quote}

In both of these sessions, only the morphological dataset \texttt{``morpho.ss''} 
is read into \poy.

\subsection{Reporting errors}
If there is an error pertaining to incorrect syntax (such as a typo in a command 
name), \poy will indicate the location of the error by underlining the problematic 
part of the input with a hat symbol (``\texttt{\^}'') in the \emph{Interactive Console} 
(Figure~\ref{fig:errors}). The description of the corresponding command, its syntax, 
and examples of its usage from the help file are automatically printed in the 
\emph{POY Output} window. As noted above, the Up and Down keys can be 
used to scroll through the output and determine the source of the error. Certain 
types of errors are reported explicitly (Figure~\ref{fig:errors}).

\begin{figure}
\centering
\begin{minipage}[c]{0.48\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figerror1.jpg}
\end{minipage}%
\,
\begin{minipage}[c]{0.48\textwidth}
\includegraphics[width=\textwidth]{doc/figures/figerror2.jpg}
\end{minipage}
\caption{Displaying errors. \poy displays error messages in several
ways. In the example in the left panel, the command \commandstyle{build}
was entered without parentheses, which is required for a  valid
\poy command syntax; the exact place of the error is marked by
``\texttt{\^}'', in this case  following the \commandstyle{build}
commands. Examples of the proper usage of the command are automatically
displayed in the \emph{POY Output}.  In other cases (right panel),
error messages are explicitly reported in the \emph{POY Output}
window. The first and second error messages indicate that the data
file \texttt{SSU.seq} is not present, which could have been caused
either by a mistake in the name of the file, or missing file, or the
location of the file in a directory, other than the one specified
prior to starting the \poy session. The third error message indicates
that the valid syntax of \commandstyle{exit} requires the parentheses
following the command name (also shown by ``\texttt{\^}'' in  the
\emph{Interactive Console}).} \label{fig:errors} \end{figure}

\subsection{Exiting} To finish a \poy session, enter the command
\commandstyle{exit()} (Figure~\ref{fig:exithelp}) or \commandstyle{quit()}.
This will close the \poy interface and resume the Terminal window
(Mac OSX) or the Command Prompt window (Windows).

\begin{figure}[]
\begin{center}
\includegraphics[width=0.7\textwidth]{doc/figures/exithelp.jpg}
\end{center}
\caption{Exiting \poy}
\label{fig:exithelp}
\end{figure}

\section{Creating and running \poy scripts}

So far, we have communicated with \poy interactively through the
\emph{Graphical User Interface} or by executing commands from the
\emph{Interactive Console}. Another way of conducting an analysis
is to run a \emph{script}, a simple text file containing a list of
commands to be performed (Figure~\ref{fig:script}).

Running analyses using scripts has many advantages: not only does
it allow for the entire analysis to proceed from the beginning to
the end at one click of a button, but it also provides means to
examine the logical dependency of the commands and optimize memory
consumption (see the description of \poyargument{script\_analysis}
argument of the command \poycommand{report} in the \emph{POY Commands}
chapter). Submitting jobs using scripts may produce results faster
because \poy automatically optimizes the workflow of the entire
analysis by taking into account the functional relationships among
various tasks and efficiently distributing the jobs and resources
(such as memory and multiple processors).

Another advantage of using scripts is that they may contain comments
that are ignored by \poy but can be helpful to describe the contents
of the files and provide other annotations. The comments are enclosed
in parenthesis \emph{and} asterisks, e.g. \texttt{(* this is a
comment *)}. Comments can be of any length and span multiple lines.

Obviously, using scripts requires the user to design the workflow
of the process prior to conducting the analysis.  \poy scripts can
be created and saved using the \emph{Script Editor} window of the
\poy \emph{Graphical User Interface} or any conventional text editor
(such as TextPad, TextWrangler, BBEdit, Emacs, or NotePad).

\poy scripts are extremely useful in cases when operations may take
a long time to complete, eliminating the need to wait for a part
of the analysis to finish in order to proceed to the next step.
Moreover, scripts can contain a series of individual scripts that 
are run sequentially (see for example tutorial \texttt{5.4}.

There are two ways to import and run a script:
\begin{itemize}
\item using the \emph{POY Launcher} in the \emph{Graphical User Interface};
\item using the command \commandstyle{run()} of the \emph{Interactive Console}; 
for example, \texttt{run("script.txt")}, where \texttt{script.txt} is the name of 
the file containing the script. Within this script it is also possible to specify 
additional scripts to run.
\end{itemize}

It is  critical to include the command \commandstyle{exit()} at the
end of the script. Otherwise \poy will be waiting for further
instructions to be entered after executing the script's contents.

\begin{figure}
\centering
\begin{minipage}[c]{0.42\textwidth}
\includegraphics[width=\textwidth]{doc/figures/commandlist.jpg}
\end{minipage}
\,
\begin{minipage}[c]{0.53\textwidth}
\includegraphics[width=\textwidth]{doc/figures/script.jpg}
\end{minipage}
\caption{Using \poy scripts. The list of commands executed interactively
using the \emph{Interactive Console} (left) and a script containing
the same list of commands (right). Observe that the header of the
script is a comment, enclosed in ``(* *)'', that is ignored by \poy.
Also note that commands can either be listed in a row or in a column
(compare \commandstyle{build()} and \commandstyle{swap()} in the
console and in the script) and different arguments of the same
command can either be specified separately or combined in a single
argument list (compare \commandstyle{report()} in the console and
in the script). (Both conventions are valid for interactive command
submission and for scripts.)} 
\label{fig:script}
\end{figure}

\section{Obtaining help} \label{sec:help} 
Instructions to run \poy, command descriptions, and the theory behind \poy can be obtained
from a variety of sources.  

\begin{description} 
\item[POY5.0 Program Documentation] (this manual) is a comprehensive 
and detailed manual on all the aspects of using \poy, from installation to 
output and visualization of results. Included are \emph{Quick Start}, \poy
command reference, practical guides and tutorials that make the
program immediately accessible for beginners and provide in-depth
information for experienced users. As with any PDF, this document 
is easily searched using key words or phrases. The documentation in PDF format
can be accessed from the \emph{Help} menu of the graphical user
interface or downloaded separately from \poy web site at \begin{center}
\url{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy/download}
\end{center} 
\item[POY] interactive help can be obtained by entering
\commandstyle{help()} at the \poy \emph{Interactive Console}.  To
obtain help on a particular command, the name of the command must
be specified in the parentheses following \commandstyle{help()}.
For example, to learn about the command \commandstyle{exit}, type
\commandstyle{help(exit)}.  (Figure~\ref{fig:exithelp}.) 
\item[POY5
Mail Group] is an Internet-based forum for discussing all issues
related to \poy and provides the best way to communicate with \poy
developers on specific issues (see \emph{WWW resources} below). The
website is located at \url{https://groups.google.com/forum/#!forum/poy4}.
Questions relating to both \texttt{POY4} and \poy can be posed to
this group.  
\item[POY Book] (Wheeler et al., 2006 \emph{Dynamic
Homology and Phylogenetic Systematics: A Unified Approach Using
POY} \cite{wheeleretal2006}) provides a review of the theory behind
\texttt{POY4} and by extension \poy, and contains formal descriptions
of many algorithms implemented in the program and the descriptions
of commands of the earlier version, \texttt{POY3}. A PDF version of 
the book is available at \url{http://research.amnh.org/scicomp/pdfs/wheeler/Wheeler_etal2006b.pdf} 
\item[POY Paper] (Var\'on et al., 2010. \emph{POY version 4: 
Phylogenetic analysis using dynamic homologies} \cite{Varonetal2010} 
provides a description of the overall goals, implementation and philosophy 
of POY.

\begin{figure}[htbp]
\centering
\includegraphics[width=0.23\textwidth]{doc/figures/figpoybook.jpg}
\caption{The POY Book.}
\label{fig:figprocess}
\end{figure}
\end{description}

\section{WWW resources}
\poy is an ongoing project and new versions are being continuously
developed to include new procedures, improve performance, and
eliminate reported bugs. Therefore, it is imperative to keep up
with the program's development and check regularly for updates.
There are several Internet-based resources that offer this information.

\begin{description}
\item[POY5 Web Site] Has downloadable compressed files of \poy
binaries, source code, and documentation in PDF format. It also
provides a links to the \emph{POY Mail Group}. The website is hosted
by AMNH Computational Sciences at \begin{center}
\url{http://www.amnh.org/our-research/computational-sciences/research/projects/systematic-biology/poy}
\end{center}

\item[POY5 source code repository] Contains has downloadable \poy
source code.  The site is powered by Google at \begin{center}
\url{http://code.google.com/p/poy/source} \end{center}

\item[POY5 Mail Group] Informs registered users via email of new
developments, such as new versions and updates.  It also provides
additional resources for obtaining help and a way for reporting
bugs and other problems with \poy and its documentation. In addition,
it allows users to receive and respond to each other's questions
thus providing an open forum to  discuss the methods and applications
of \poy. The users who choose not to register, have access to the
archives of the postings but will not be able to either submit or
receive emails from other users and \poy developers. The \emph{POY5
Mail Group} is hosted  by Google at \begin{center}
\url{https://groups.google.com/forum/#!forum/poy4} \end{center}

\end{description}